src/Microsoft.ML.OnnxTransform/OnnxTransform.cs

```diff
@@ -333,91 +352,103 @@ protected override Delegate MakeGetter(IRow input, int iinfo, Func<int, bool> ac
+            private static INamedOnnxValueGetter CreateNameOnnxValueGetter<T>(IRow input, bool isVector, string colName, int colIndex, OnnxShape onnxShape)
             {
                 if (isVector)
-                    return new TensorValueGetterVec<T>(input, colIndex, onnxShape);
-                return new TensorValueGetter<T>(input, colIndex);
+                    return new NamedOnnxValueGetterVec<T>(input, colName, colIndex, onnxShape);
+                return new NameOnnxValueGetter<T>(input, colName, colIndex);
             }
 
-            private class TensorValueGetter<T> : ITensorValueGetter
+            private class NameOnnxValueGetter<T> : INamedOnnxValueGetter
             {
                 private readonly ValueGetter<T> _srcgetter;
+                private readonly string _colName;
 
-                public TensorValueGetter(IRow input, int colIndex)
+                public NameOnnxValueGetter(IRow input, string colName, int colIndex)
                 {
+                    _colName = colName;
                     _srcgetter = input.GetGetter<T>(colIndex);
                 }
-                public Tensor GetTensor()
+                public NamedOnnxValue GetNamedOnnxValue()
                 {
                     var scalar = default(T);
                     _srcgetter(ref scalar);
-                    return OnnxUtils.CreateScalarTensor(scalar);
+                    return OnnxUtils.CreateScalarNamedOnnxValue(_colName, scalar);
                 }
             }
 
-            private class TensorValueGetterVec<T> : ITensorValueGetter
+            private class NamedOnnxValueGetterVec<T> : INamedOnnxValueGetter
             {
                 private readonly ValueGetter<VBuffer<T>> _srcgetter;
                 private readonly OnnxShape _tensorShape;
+                private readonly string _colName;
                 private VBuffer<T> _vBuffer;
                 private VBuffer<T> _vBufferDense;
-                public TensorValueGetterVec(IRow input, int colIndex, OnnxShape tensorShape)
+                public NamedOnnxValueGetterVec(IRow input, string colName, int colIndex, OnnxShape tensorShape)
                 {
                     _srcgetter = input.GetGetter<VBuffer<T>>(colIndex);
                     _tensorShape = tensorShape;
+                    _colName = colName;
                     _vBuffer = default;
                     _vBufferDense = default;
                 }
-                public Tensor GetTensor()
+                public NamedOnnxValue GetNamedOnnxValue()
                 {
                     _srcgetter(ref _vBuffer);
                     _vBuffer.CopyToDense(ref _vBufferDense);
-                    return OnnxUtils.CreateTensor(_vBufferDense.GetValues(), _tensorShape);
+                    return OnnxUtils.CreateNamedOnnxValue(_colName, _vBufferDense.GetValues(), _tensorShape);
```

Comment:
It's unfortunate how many copies we are making here.

1.`_srcgetter` will copy the data into `_vBuffer`.
2. `_vBuffer` will copy the dense data into `_vBufferDense`
3. `OnnxUtils.CreateNamedOnnxValue` calls `ToArray()` on the Span passed into it, which will make a 3rd copy.

I think in a "perfect" world, ML.NET would have "Tensor" as a "well-known" type, just like `VBuffer`. Then we could talk in terms of `Tensor` here instead of `VBuffer`, and not need to do so many copies.

Reply:
Point noted. Currently the APIs require this level of copying. We can track this separately. I'll close for now.

---
In reply to: [236341262](https://github.com/dotnet/machinelearning/pull/1717#discussion_r236341262) [](ancestors = 236341262)